File: ruby/Lexer.rb
require './Token.rb'

class Lexer

  attr_accessor :code, :tokens, :token_string, :current_c

  def initialize
    @code = ""
    @tokens = []
    @token_string = ""
    @current_c = ""
    @line = 1
  end

  def run(code, debug: false)
    initialize
    @code = code
    capturing_string = false

    @code.split('').each do |c|
      @current_c = c

      case c
      when "\""
        @token_string << c
        lock_token(Type::STRING_LIT) if capturing_string
        capturing_string = !capturing_string

      else
        if capturing_string
          @token_string << c
        else
          case c
          when "\n", " ", "\t"
            label_token if @token_string.size != 0
            @token_string << c if capturing_string

          when "("
            id_locker(Type::LPAR)

          when ")"
            id_locker(Type::RPAR)

          when ","
            id_locker(Type::COMMA)

          when ";"
            id_locker(Type::SEMI)

          when '{'
            id_locker(Type::LBRACE)

          when '}'
            id_locker(Type::RBRACE)

          when '.'
            id_locker(Type::DOT)

          else
            @token_string << c
          end
        end
      end

      @line += 1 if c == "\n"
    end

    label_token if @token_string.size != 0

    @tokens << Token.new(scope: Scope::LEXING, type: Type::FINISH_PARSING, value: "$", line: @line, id: tokens.size)

    print_tokens if debug

    return @tokens
  end

  def lock_token(type)
    if @token_string.size != 0
      @tokens << Token.new(scope: Scope::LEXING, type: type, value: @token_string, line: @line, id: tokens.size)
      @token_string = ""
    end
  end

  def label_token
    case @token_string
    when /void/
      lock_token(Type::VOID)

    when /i32/
      lock_token(Type::I32)

    when /string/
      lock_token(Type::STRING)

    when /[0-9]+/
      lock_token(Type::INT_LIT)

    when /end/
      lock_token(Type::END_)

    when /do/
      lock_token(Type::DO)

    when /return/
      lock_token(Type::RETURN)

    when /bind/
      lock_token(Type::BIND)

    when /if/
      lock_token(Type::IF)

    when /else/
      lock_token(Type::ELSE)

    when /</
      lock_token(Type::LT)

    when /\"[^\"]*\"/
      lock_token(Type::STRING_LIT)

    else
      lock_token(Type::ID)
    end
  end

  def id_locker(type)
    lock_token(Type::ID)
    @token_string << @current_c
    lock_token(type)
  end

  def print_tokens
    puts "\n--- TOKENS ---\n"
    @tokens.each do |token|
      puts token.to_s
    end
  end

end

File: lemon/lexer.cpp
#include "lexer.hpp"

void Lexer::lock_token(type_t type) {
  if (token_string.size() == 0) return;
  tokens.push_back(token_t(scope_t::LEXER, type, token_string, line, tokens.size()));
  token_string = "";
}

void Lexer::label_token() {
  if (regex_match(token_string, regex("[0-9]+"))) {
    lock_token(type_t::INT_LIT);
  } else if (regex_match(token_string, regex("void"))) {
    lock_token(type_t::VOID);
  } else if (regex_match(token_string, regex("i32"))) {
    lock_token(type_t::I32);
  } else if (regex_match(token_string, regex("string"))) {
    lock_token(type_t::STRING);
  } else if (regex_match(token_string, regex("end"))) {
    lock_token(type_t::END);
  } else if (regex_match(token_string, regex("do"))) {
    lock_token(type_t::DO);
  } else if (regex_match(token_string, regex("return"))) {
    lock_token(type_t::RETURN);
  } else if (regex_match(token_string, regex("bind"))) {
    lock_token(type_t::BIND);
  } else if (regex_match(token_string, regex("if"))) {
    lock_token(type_t::IF);
  } else if (regex_match(token_string, regex("else"))) {
    lock_token(type_t::ELSE);
  } else if (regex_match(token_string, regex("<"))) {
    lock_token(type_t::LT);
  } else if (regex_match(token_string, regex("\"[^\"]*\""))) {
    lock_token(type_t::STRING_LIT);
  } else {
    lock_token(type_t::ID);
  }
}

void Lexer::id_locker(type_t type) {
  lock_token(type_t::ID);
  token_string.push_back(*source_it);
  lock_token(type);
}

list<token_t> Lexer::run() {
  bool capturing_string = false;
  source_it = source.begin();

  while (source_it != source.end()) {
    switch (*source_it) {
      case '"':
        token_string.push_back(*source_it);
        if (capturing_string) lock_token(type_t::STRING);
        capturing_string = !capturing_string;
        break;

      /* spaces */
      case '\n':
      case ' ':
      case '\t':
        if (!capturing_string && token_string.size() != 0) {
          token_it = token_string.begin();
          label_token();
        }

        if (capturing_string) token_string.push_back(*source_it);
        break;

      default:
        switch (*source_it) {
          case '(':
            id_locker(type_t::LPAR);
            break;

          case ')':
            id_locker(type_t::RPAR);
            break;

          case ',':
            id_locker(type_t::COMMA);
            break;

          case ';':
            id_locker(type_t::SEMI);
            break;

          case '{':
            id_locker(type_t::LBRACE);
            break;

          case '}':
            id_locker(type_t::RBRACE);
            break;

          case '.':
            id_locker(type_t::DOT);
            break;

          default:
            token_string.push_back(*source_it);
        }
    }

    if (*source_it == '\n') ++line;

    ++source_it;
  }

  if (token_string.size() != 0) {
    token_it = token_string.begin();
    label_token();
  }

  tokens.push_back(token_t(scope_t::LEXER, type_t::FINISH_PARSING, "", line, tokens.size()));

  return tokens;
}


void Lexer::print_tokens() {
  cout << endl << "--- TOKENS ---" << endl;
  for (auto token : tokens) {
    cout << token.type_to_string() << "(" << token.id << ", " << token.line  << "): " << token.value << endl;
  }
}